sim,python: Restore sigint handler in python

powerjg opened the pull request:
Currently, if you try to use ctrl-c while python code is running nothing
happens. This is not ideal. This change enables users to use ctrl-c
while python is running (e.g., when a large disk image is downloading).
To do this, we moved the `initSignals` function in gem5 from `main` to
the simulate loop. Thus, every time the simulate loop starts (i.e., is
called from python) gem5 will install its signal handlers. Also, when
the control is returned to python, we put python's default SIGINT
handler back.

Change-Id: I14490e48d931eb316e8c641217bf8d8ddaa340ed
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg commented:
@andysan here's another attempt at the signal handler problem. Let me know what you think.

andysan commented:
At a high-level, I think this sort of makes sense. It is slightly weird that you can't interrupt the Python code when you are not executing the event loop. From an implementation perspective, I think it would be better to save the old SIGINT handler using `sigaction` when you install gem5's SIGINT handler in simulate, and then restore it when returning. That avoids the fixup code in Python and makes the behaviour consistent if built without Python. That should make the code easier to follow since the intent is clearer.

If you want to make this really Python friendly, we should consider throwing a `KeyboardInterrupt` from `m5.simulate` if the simulator was interrupted. This would definitely break some scripts, so we need to be careful if we do this.

powerjg commented:
> I think it would be better to save the old SIGINT handler using sigaction when you install gem5's SIGINT handler in simulate, and then restore it when returning. 

Great idea... I don't know why I didn't do it that way...

> If you want to make this really Python friendly, we should consider throwing a KeyboardInterrupt from m5.simulate if the simulator was interrupted. This would definitely break some scripts, so we need to be careful if we do this.

I think this is a bigger change than what we want to do right now. Personally, I think that this should be tied to a better interface for m5 exit events. This is on our short-term to do list, so hopefully we'll tackle this soon. See #474

powerjg commented:
Thanks for the feedback, Andreas. I've implemented this as you suggested. I've tested it by hitting "ctrl-c" while gem5 is running the sim loop and during the python execution before and after the sim loop. That now works as expected. I have not tested other signals, but I don't have any reason they should behave differently.

andysan commented:
Thanks, this looks a lot better! However, I'm a bit worried that restoring the other signal handlers might lead to weird behaviour. For example, if you restore SIGUSRx, you could end up in a situation where a stat dump isn't triggered if you happen to be servicing an event in Python. That seems wrong to me. Similarly, we probably want to print a backtrace if a call from Python triggers an abort or segfault in the C++ code. For the problem you're solving, I think it would be sufficient to just save and restore SIGINT.

BobbyRBruce commented:
@powerjg : Are you wanting this to get in for v23.1? If so, do you need help moving this forward?

powerjg commented:
There's little chance I'll have time until after the ASPLOS deadline. I looked at Andreas's suggestion and it's not totally straightforward.

andysan commented:
> There's little chance I'll have time until after the ASPLOS deadline. I looked at Andreas's suggestion and it's not totally straightforward.

As far as I can tell, you'd get the desired behaviour if you just save and restore SIGINT. It seems to me like the other signal handlers can be left in place when executing Python code, That way, we can still intercept SIGUSRx for stat dumps, and SIGSEGV and SIGABRT for backtraces.

powerjg commented:
I understand your suggestion, but I don't have time to implement it right now, unfortunately. I appreciate the review, but often times things that looks like a 5 minute job to the reviewer take significantly more effort.

The problem is that now we need two separate functions to initialize the signals with your suggestion. It's not hard, but it's not a 5 minute job. I estimate 30-60 minutes + testing. I just don't have the time right now with deadlines, etc.

So, it looks like this is going to languish unless someone else picks it up.

powerjg committed:
sim,python: Restore sigint handler in python

Currently, if you try to use ctrl-c while python code is running nothing
happens. This is not ideal. This change enables users to use ctrl-c
while python is running (e.g., when a large disk image is downloading).
To do this, we add a new function that only initializes the sigint
handler and call that just before the event loop begins processing.
Then, when the event loop is done, we reinstall the original handler.

Change-Id: I14490e48d931eb316e8c641217bf8d8ddaa340ed
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg commented:
Ok, I've tried a different approach. Let me know if this is what you were thinking.

That said, I don't like that we're treating SIGINT uniquely. This code could easily get out of hand.

andysan commented:
> Ok, I've tried a different approach. Let me know if this is what you were thinking.
> 
> That said, I don't like that we're treating SIGINT uniquely. This code could easily get out of hand.

I think this this is fine. The other signals need to be handled inside gem5 itself, SIGINT is the only one that needs to hand over control to the simulation script.
